src/meta/client/MetaClient.cpp

```diff
@@ -897,6 +916,115 @@ MetaClient::dropEdgeSchema(GraphSpaceID spaceId, std::string name) {
     }, true);
 }
 
+folly::Future<StatusOr<TagIndexID>>
+MetaClient::createTagIndex(GraphSpaceID spaceID, std::string name,
+                           std::map<std::string, std::vector<std::string>> fields) {
+    cpp2::IndexProperties properties;
+    properties.set_fields(std::move(fields));
+    cpp2::CreateTagIndexReq req;
+    req.set_space_id(std::move(spaceID));
+    req.set_index_name(std::move(name));
+    req.set_properties(std::move(properties));
+
+    return getResponse(std::move(req), [] (auto client, auto request) {
+        return client->future_createTagIndex(request);
+    }, [] (cpp2::ExecResp&& resp) -> TagIndexID {
+        return resp.get_id().get_tag_index_id();
+    }, true);
+}
+
+folly::Future<StatusOr<bool>>
+MetaClient::dropTagIndex(GraphSpaceID spaceID, std::string name) {
+    cpp2::DropTagIndexReq req;
+    req.set_space_id(std::move(spaceID));
+    req.set_index_name(std::move(name));
+
+    return getResponse(std::move(req), [] (auto client, auto request) {
+        return client->future_dropTagIndex(request);
+    }, [] (cpp2::ExecResp&& resp) -> TagIndexID {
+        return resp.get_id().get_tag_index_id();
+    }, true);
+}
+
+folly::Future<StatusOr<cpp2::TagIndexItem>>
+MetaClient::getTagIndex(GraphSpaceID spaceID, std::string name) {
+    cpp2::GetTagIndexReq req;
+    req.set_space_id(std::move(spaceID));
+    req.set_index_name(std::move(name));
+
+    return getResponse(std::move(req), [] (auto client, auto request) {
+        return client->future_getTagIndex(request);
+    }, [] (cpp2::GetTagIndexResp&& resp) -> nebula::meta::cpp2::TagIndexItem {
+        return std::move(resp).get_item();
+    }, true);
+}
+
+folly::Future<StatusOr<std::vector<cpp2::TagIndexItem>>>
+MetaClient::listTagIndexes(GraphSpaceID spaceID) {
+    cpp2::ListTagIndexesReq req;
+    req.set_space_id(std::move(spaceID));
+
+    return getResponse(std::move(req), [] (auto client, auto request) {
+        return client->future_listTagIndexes(request);
+    }, [] (cpp2::ListTagIndexesResp&& resp) -> decltype(auto) {
+        return std::move(resp).get_items();
+    });
+}
+
+folly::Future<StatusOr<EdgeIndexID>>
+MetaClient::createEdgeIndex(GraphSpaceID spaceID, std::string name,
+                            std::map<std::string, std::vector<std::string>> fields) {
```

Comment:
because here use fields to construct the request, so I think fields need &&.

Reply:
Thanks for you great idea :)

It's seems will be a huge task and I will open another PR to process it .

Reply:
After talk with `wadeliuyi` offline, it's a big task and we should open another PR to process it. thanks